### PR TITLE
concatenate polylines

### PR DESCRIPTION
**Pull request by haoliangyu**
This function concatenates polylines if their start/end points are within the specific tolerance. It's not equal to #5 since the function doesn't care about any intermediate vertices, no matter how close they are. The primary use case of this function is to rebuild a complete polyline from an array of line segments.The output is an array in case some polylines are unable to be concatenated into the other.

It works well in my use case and I have designed some test cases for it. Not sure whether they covers all cases, so please let me know if I can make improvement.

### CODE REVIEW

File: dist/leaflet.geometryutil.js [hunk 1]
```diff
@@ -498,6 +498,78 @@ L.GeometryUtil = L.extend(L.GeometryUtil || {}, {
     },
 
     /**
+     * Concatenate polylines whose start/end points are within a close distance
+     * @param  {L.Map}        map
+     * @param  {L.Polyline[]} polylines         An array of L.Polyline objects
+     * @param  {Number}       [tolerance=15]    A pixel distance in which points will be considered equal
+     * @return {L.Polyline[]} concateLines      An array of concatenated L.Polyline objects
+     */
+    concatLines: function(map, polylines, tolerance) {
```
Comment (leplatrem) [inline dist/leaflet.geometryutil.js]: nitpick : in PostGIS, it is called *line merge*. I suggest we rename it to mimic the standard. What do you think ?

Comment (leplatrem) [inline dist/leaflet.geometryutil.js]: Maybe it matches the *union* notion actually.

File: test/test.geometryutil.js [hunk 2]
```diff
@@ -477,6 +477,53 @@ describe('Line order', function() {
...
+  });
+
+  it('It should return the original array if there is one given polyline', function(done) {
+    var line = L.polyline([[0, 0], [1, 1]]);
+    var concated = L.GeometryUtil.concatLines(map, [line]);
+    assert.equal(concated.length, 1);
+    done();
+  });
+
+  it('It should return an arry of concatenated lines', function(done) {
```
Comment (leplatrem) [inline test/test.geometryutil.js]: typo: array

File: dist/leaflet.geometryutil.js (hunk 1 above)
Comment (leplatrem) [inline dist/leaflet.geometryutil.js]: It reminds me this :) https://github.com/makinacorpus/Geotrek/blob/master/geotrek/core/tests/test_path_merge.py#L12 /cc @gutard @fredericbonifas

File: test/test.geometryutil.js [hunk 3]
```diff
@@ -477,6 +477,53 @@ describe('Line order', function() {
...
+    assert.isTrue(concatedLatLngs[0].equals(L.latLng([35, 35])));
+    assert.isTrue(concatedLatLngs[3].equals(L.latLng([20, 20])));
+
+    var concatedLatLngs = concated[1].getLatLngs();
+    assert.equal(concatedLatLngs.length, 3);
+    assert.isTrue(concatedLatLngs[0].equals(L.latLng([2, 1])));
+    assert.isTrue(concatedLatLngs[2].equals(L.latLng([1, 1])));
+
+    done();
+  })
```
Comment (leplatrem) [inline test/test.geometryutil.js]: I suggest that you add a couple of additional tests (simple ones with two lines for example), for the following cases:

* tolerance is 0
* the input paths are not modified
* the two paths form a «\/» or a /\»
* the two paths are equal
* the two paths are equal but opposite direction
* the two paths form a loop

**Comment by leplatrem** [comment]
This is a very good contribution!

If only we only had contributors like you that write tests for the code they propose!

Thank you very much for your efforts on this!

File: dist/leaflet.geometryutil.js (hunk 1 above)
Comment (haoliangyu) [inline dist/leaflet.geometryutil.js]: I think `lineMerge()` will be a good name for this specific functionality. The meaning of `union` seems to be too broad.

File: test/test.geometryutil.js (hunk 3 above)
Comment (haoliangyu) [inline test/test.geometryutil.js]: It makes sense to do more testing :)

**Comment by haoliangyu** [comment]
PR closes for additional work.

**Pull request closed without merging**